portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```
@@ -26,24 +33,83 @@
+  private static final String COMMENT_NAME_PARAM = "name";
+  private static final String COMMENT_NUM_PARAM = "num-comments";
+  private static final String ENTITY_QUERY = "Comment";
+  private static final String ENTITY_TEXT_PARAM = "comment-text";
+  private static final String ENTITY_NAME_PARAM = "username";
+  private static final String ENTITY_TIME_PARAM = "time-posted";
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   @Override
-  public void init() {
-    names = new ArrayList<>();
-    names.add("Cindy");
-    names.add("Anthony");
-    names.add("Charles");
-    names.add("Chris");
-    names.add("Ben");
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String comment = request.getParameter(COMMENT_TEXT_PARAM);
+    String name = request.getParameter(COMMENT_NAME_PARAM);
+
+    long timestamp = System.currentTimeMillis();
+
+    if (name.equals("")) {
+      name = "Anonymous";
+    }
+
+    Entity commentEntity = new Entity(ENTITY_QUERY);
+    commentEntity.setProperty(ENTITY_TEXT_PARAM, comment);
+    commentEntity.setProperty(ENTITY_NAME_PARAM, name);
+    commentEntity.setProperty(ENTITY_TIME_PARAM, timestamp);
+
+    datastore.put(commentEntity);
+
+    response.sendRedirect("/comments.html");
   }
+
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     Gson gson = new Gson();
-    String json = gson.toJson(names);
+    int num_comments;
+    String user_num = request.getParameter(COMMENT_NUM_PARAM);
+
+    try {
+      num_comments = Integer.parseInt(user_num);
+    } catch (NumberFormatException e) {
+      System.err.println("Could not convert to int: " + user_num);
+      num_comments = -1;
+    }
+
+    if (num_comments < 0 || num_comments > 15) {
+      String json = gson.toJson("Please enter an integer between 0 and 15.");
```

This is a good intuition. Rather than silently return a JSON response (which could be interpreted as a normal response), it's better to return an HTTP error code. Since you're already using a JSON coder, you could even return a structured error response, e.g., `{"type": "VALIDATION", "message": "Please enter an integer between 0 and 15"}` (if it makes sense to put errors into different classes at least).

To set a response code:

```
response.setStatus(HttpServletResponse.SC_BAD_REQUEST); // 404 -- bad request
response.setContentType(..);
```

NOTE: You may need to set the response type _before_ writing anything to the response writer. Some frameworks assume a 200 response code if you start writing before setting status codes.